### ISSUE TITLE
How to get the grandparent data in one resolver ?

### ISSUE THREAD

**Issue by tychota** [graphql/graphql-js#1098]
-- EDIT --

We can return whatever data we want in resolver and thus pass the competition deeper down the tree. Fixed.

-- EDIT --

*This issue is more a question than a feature request or a bug.*
*I'm not sure what are the rules in this repository about "stackoverflow" issues.* 

*That being said, it seems a pretty much unanswered problem across internet so:*
- *I can commit on updating the doc with a FAQ about this once it is solved.*
- *I can help build a utils function in GraphQL to help solve this usecase*

## Context

I'm building a GraphQL API, as a proxy to a existing, hard to modify REST API.

I'm not in charge of this REST API, so I would consider it isn't possible to change/adapt the REST API.

I have a data graph that look like this:

<p align="center">
  <img src="https://user-images.githubusercontent.com/13785185/33234005-9f2fef70-d21f-11e7-9d2a-134bf935f9b2.png" height="50%" width="50%" />
</p>

I'm wanting to be able to perform this GraphQL query:

<p align="center">
  <img src="https://user-images.githubusercontent.com/13785185/33234036-06a6cdea-d220-11e7-922a-03d9b293de6c.png" height="50%" width="50%" />
</p>

### Types

<details> 
  <summary><b>Query.graphql</b></summary>

```gql
type Query {}
```

</details>

<details> 
  <summary><b>Team.graphql</b></summary>

```gql
type Team {
  displayName: String!
  id: ID!
  # ... other properties
}

extend type Query {
  team(id: ID!): Team
}
```

</details>

<details> 
  <summary><b>Competition.graphql</b></summary>

```gql
type Competition {
  displayName: String!
  id: ID!
  season(id: ID!): Season
  seasons: [Season]
  # ... other properties
}

extend type Query {
  competitions: [Competition]
  competition(id: ID!): Competition
}
```

</details>

<details> 
  <summary><b>Competition.graphql</b></summary>

```gql
type Competition {
  displayName: String!
  id: ID!
  season(id: ID!): Season
  seasons: [Season]
  # ... other properties
}

extend type Query {
  competitions: [Competition]
  competition(id: ID!): Competition
}
```

</details>

<details> 
  <summary><b>Season.graphql</b></summary>

```gql
enum SeasonStatus {
  ACTIVE
  INACTIVE
}

type Season {
  id: ID!
  seasonYear: Int!
  status: SeasonStatus
  teams: [Team]
  team(id: ID!): Team
  winner: Team
   # ... other properties
}
```

</details>

### Resolvers

<details> 
  <summary>✅ <b>query/competitions.js</b></summary>

```javascript
import fetch from "../../utils/fetch";
import {ROOT_URL} from "../../config";

export default async (obj, args, context, info) => {
  const response = await fetch(
    `${ROOT_URL}/v1/competitions`
  );
  if (!response.ok) throw new Error(response.status);
  const data = await response.json();
  return data;
};
```

</details>

<details> 
  <summary>✅ <b>query/competition.js</b></summary>

```javascript
import fetch from "../../utils/fetch";
import {ROOT_URL} from "../../config";

export default async (obj, args, context, info) => {
  const response = await fetch(
    `${ROOT_URL}/v1/competitions/${args.id}`
  );
  if (!response.ok) throw new Error(response.status);
  const data = await response.json();
  return data;
};
```

</details>

<details> 
  <summary>✅ <b>competition/seasons.js</b></summary>

```javascript
import fetch from "../../utils/fetch";
import {ROOT_URL} from "../../config";

export default async (obj, args, context, info) => {
  const response = await fetch(
    `${ROOT_URL}/v1/competitions/${obj.id}/seasons`
  );
  if (!response.ok) throw new Error(response.status);
  const data = await response.json();
  return data;
};
```

</details>

<details> 
  <summary>✅ <b>competition/season.js</b></summary>

```javascript
import fetch from "../../utils/fetch";
import {ROOT_URL} from "../../config";

export default async (obj, args, context, info) => {
  const response = await fetch(
    `${ROOT_URL}/v1/competitions/${obj.id}/seasons/${args.id}`
  );
  if (!response.ok) throw new Error(response.status);
  const data = await response.json();
  return data;
};
```

</details>

<details> 
  <summary>⛔️ <b>season/team.js</b></summary>

```javascript
import fetch from "../../utils/fetch";
import {ROOT_URL} from "../../config";

export default async (obj, args, context, info) => {
  let competitionId = "How to get that ??";
  let seasonId = obj.id;
  let teamId = args.id;
  const response = await fetch(
    `${ROOT_URL}/v1/competitions/${competitionId}/seasons/${seasonId}/teams/${teamId}`
  );
  if (!response.ok) throw new Error(response.status);
  const data = await response.json();
  return data;
};
```

</details>

## My problem

<p align="center">
  <img src="https://cdn.discordapp.com/attachments/103315563073269760/384017087829639169/unknown.png" height="50%" width="50%" />
</p>

Retrieving the green part work.

**I don't manage retrieving the red part.** 😭

That is because for the red part I have to make the following resolver:

```js
import fetch from "../../utils/fetch";
import {ROOT_URL} from "../../config";

export default async (obj, args, context, info) => {
  let competitionId = "How to get that ??";
  let seasonId = obj.id;
  let teamId = args.id;
  const response = await fetch(
    `${ROOT_URL}/v1/competitions/${competitionId}/seasons/${seasonId}/teams/${teamId}`
  );
  if (!response.ok) throw new Error(response.status);
  const data = await response.json();
  return data;
};
```

## My problem analysis

I have no easy way to access the grandparent data.

### 1. Restructuring the schema to flatten it

<p align="center">
  <img src="https://user-images.githubusercontent.com/13785185/33234236-65300644-d223-11e7-859c-f1ad2bb33c8f.png" height="50%" width="50%" />
</p>

I did not like this solution because it don't fit the domain well.

It makes `Competition` type disappear, and force some renaming:
- `competition.displayName` will become `season.competionName`

### 2. Add a competitionId field in season

It may be ok in this case but still seems like an ugly hack.

I can modify those files like this:

- `Season` type
  ```gql
  type Season {
    id: ID!
    seasonYear: Int!
    status: SeasonStatus
    teams: [Team]
    team(id: ID!): Team
    winner: Team
     # ... other properties
    competitionID: ID
  }
  ```

- `competition/season.js` resolver:
  ```js
  import fetch from "../../utils/fetch";
  import {ROOT_URL} from "../../config";
  
  export default async (obj, args, context, info) => {
    const response = await fetch(
      `${ROOT_URL}/v1/competitions/${obj.id}/seasons/${args.id}`
    );
    if (!response.ok) throw new Error(response.status);
    const data = await response.json();
    return {...data, competionId: obj.id};
  };
  ```

- something similar for `competition/seasons.js` resolver

- then the team resolver would be:
  ```js
  import fetch from "../../utils/fetch";
  import {ROOT_URL} from "../../config";
  
  export default async (obj, args, context, info) => {
    let competitionId = obj.competionId;
    let seasonId = obj.id;
    let teamId = args.id;
    const response = await fetch(
      `${ROOT_URL}/v1/competitions/${competitionId}/seasons/${seasonId}/teams/${teamId}`
    );
    if (!response.ok) throw new Error(response.status);
    const data = await response.json();
    return data;
  };
  ```

### 3. Accessing grandparent from resolver

I did not find an easy way to do it.

I suppose it is doable by parsing the AST in the info argument.

At least it is what people does in Graphene: see https://github.com/graphql-python/graphene/issues/348#issuecomment-259599683

This PR did modify the data we can get in field AST so I'm a bit lost.

## Expectations

I would expect a clearer documentation about how to solve this case.

I would maybe expect an utility function that work like this.

```js

import { buildGenealogy } from "graphql";

export default async (args, context, info) => {
  const genealogy = buildGenealogy(info);

  // usage
  const parentValue = genealogy.parent.value;
  const parentArgs = genealogy.parent.args;
  const grandParentValue = genealogy.parent.parent.value;
  const isGrandParentRoot = genealogy.parent.parent.isRoot;

  // rest of the resolver
}
```

I'm willing to contribute, to doc or to this function, with some guidance.

**Comment by IvanGoncharov** [comment]
@tychota You can inject your own data into rootValue, like that:

**competition/season.js**
```ts
import fetch from "../../utils/fetch";
import {ROOT_URL} from "../../config";

export default async (obj, args, context, info) => {
  const response = await fetch(
    `${ROOT_URL}/v1/competitions/${obj.id}/seasons/${args.id}`
  );
  if (!response.ok) throw new Error(response.status);
  const data = await response.json();
  return {
     ...data,
     competitionId: obj.id
  };
};
```

**Comment by tychota** [comment]
Look like my solution 2. ^^

I did not need to modify the `Season` type to add a competitionId field ?
If so that would be really better than the solution 2 !

**Comment by IvanGoncharov** [comment]
@tychota 

TL;DR; You can use `rootValue` to pass anything you want 😃 

`rootValue`(first arg of resolver func) can be arbitrary data it's a way to pass context between resolvers and it doesn't relate to your types in any way.

I suspect you are confused because in your code you are using default behavior of field resolvers which is to grab the value of the property with the same name as the field from the `rootValue`. 
So resolver functions returning context to other resolvers (including default ones) and not a value of the instance.

**Comment by tychota** [comment]
@IvanGoncharov : working fine with you advices. 

Thank you Sir !

**Comment by IvanGoncharov** [comment]
@tychota My pleasure.

BTW. You can do a lot more if you override default resolvers. For example, you can rename fields or change it's type, etc.

> I can commit on updating the doc with a FAQ about this once it is solved.

It's a pretty common mistake to confuse `rootValue` and instance's data so having this difference described somewhere would be really helpful. 

You can expend [this section](http://graphql.org/learn/execution/#root-fields-resolvers) of official docs. It's source located here: https://github.com/graphql/graphql.github.io/edit/source/site/learn/Learn-Execution.md

Alternatively, you write a Medium article or something similar.

**Comment by tychota** [comment]
@IvanGoncharov : https://github.com/graphql/graphql.github.io/pull/425/files

Is it clear enough ??

**Comment by IvanGoncharov** [comment]
@tychota Wow you are fast 👍 
Looks good for me but I'm not a native speaker 😄 
If this PR stuck for more than a couple days just ping a few of [top contributors](https://github.com/graphql/graphql.github.io/graphs/contributors) all of them have commit rights and can merge you PR.

**Comment by rsambasivans** [comment]
Hello @IvanGoncharov and @tychota ,

I know this is much later, but I am having a similar issue that I was hoping you could answer.. My schema is similar to tychota's, but for competition/season.js, I need to return Season[] rather than Season. Since I need to return an array, it does not allow me to return in the way you suggested:

> return {
>      ...data,
>      competitionId: obj.id
>   };

When I attempt this, I get an error: Expected Iterable, but did not find one for field "<my_field>".

How can I pass additional values through rootValue when I need to return an array?

**Comment by desiprisg** [comment]
@rsambasivans Hey I know I'm late as well. You should pass an array with every object mutated. So if you have the seasons of type Season[] you should return: 
```
result.map((season) => ({ ...season, competitionId: obj.id }))
```
Think of it like this: Every item of this array will be given as a root to a different subset of teams(each subset has one root)
